src/repr/src/stats.rs

```diff
@@ -0,0 +1,49 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::{Datum, Row, RowArena};
+
+/// Provides access to statistics about SourceData stored in a Persist part (S3
+/// data blob).
+///
+/// Statistics are best-effort, and individual stats may be omitted at any time,
+/// e.g. if persist cannot determine them accurately, if the values are too
+/// large to store in Consensus, if the statistics data is larger than the part,
+/// if we wrote the data before we started collecting statistics, etc.
+pub trait PersistSourceDataStats: std::fmt::Debug {
+    /// The number of updates (Rows + errors) in the part.
+    fn len(&self) -> Option<usize>;
+
+    /// The number of errors in the part.
+    fn err_count(&self) -> Option<usize>;
+
+    /// The part's minimum value for the named column, if available.
+    /// A return value of `None` indicates that Persist did not / was
+    /// not able to calculate a minimum for this column.
+    fn col_min<'a>(&self, name: &str, arena: &'a RowArena) -> Option<Datum<'a>>;
+
+    /// (ditto above, but for the maximum column value)
+    fn col_max<'a>(&self, name: &str, arena: &'a RowArena) -> Option<Datum<'a>>;
+
+    /// The part's null count for the named column, if available. A
+    /// return value of `None` indicates that Persist did not / was
+    /// not able to calculate the null count for this column.
+    fn col_null_count(&self, name: &str) -> Option<usize>;
+
+    /// A prefix of column values for the minimum Row in the part. A
+    /// return of `None` indicates that Persist did not / was not able
+    /// to calculate the minimum row. A `Some(usize)` indicates how many
+    /// columns are in the prefix. The prefix may be less than the full
+    /// row if persist cannot determine/store an individual column, for
+    /// the same reasons that `col_min`/`col_max` may omit values.
+    fn row_min(&self, row: &mut Row) -> Option<usize>;
```

Comment:
I don't understand how this and `row_max` are going to provide added value if we already have `col_min` and `col_max`. Can you give an example?

Reply:
IAND (I am not Dan) but row min/max are useful within persist for reasons separate from MFP pushdown, like knowing the precise bounds of batches and parts to efficiently merge sort them together during compaction. Regardless of whether they are useful for filter pushdown, we'd expect to add these in for ourselves as part of the overall stats interface.

Reply:
Thanks, the API makes more sense now. I don't understand all the details of the `(3, 'z')` - `(1, 'a')` example but this doesn't seem necessary for this PR.

Reply:
Got it. I think the API will be more flexible if you can track statistics / metadata for arbitrary subsets of columns in the schema (similar to how some database will collect metadata such as `min` and `max` for `(a, b)`  whenever you create an index on `(a, b)`).